crypto/x509/v3_conf.c

```diff
@@ -365,17 +365,13 @@ int X509V3_EXT_CRL_add_nconf(CONF *conf, X509V3_CTX *ctx, const char *section,
 int X509V3_EXT_REQ_add_nconf(CONF *conf, X509V3_CTX *ctx, const char *section,
                              X509_REQ *req)
 {
-    STACK_OF(X509_EXTENSION) *extlist = NULL, **sk = NULL;
-    int i;
+    STACK_OF(X509_EXTENSION) *exts = NULL;
+    int ret = X509V3_EXT_add_nconf_sk(conf, ctx, section, &exts);
 
-    if (req)
-        sk = &extlist;
-    i = X509V3_EXT_add_nconf_sk(conf, ctx, section, sk);
-    if (!i || !sk)
-        return i;
-    i = X509_REQ_add_extensions(req, extlist);
-    sk_X509_EXTENSION_pop_free(extlist, X509_EXTENSION_free);
-    return i;
```

Comment:
What is the supposed mem leak that is being fixed here? I do not see it. Also this breaks API as before the change X509V3_EXT_REQ_add_nconf was possible to be called with NULL req to just check the extensions without adding them to X509_REQ.

Reply:
I had experienced a mem leak on failure of calling `X509V3_EXT_add_nconf_sk(conf, ctx, section, sk)`.
In this case the `sk` may be non-NULL and needs to be deallocated.

I was not aware of the 'just checking' use case; i've re-enabled it and added it to the documenting comment.
Moreover, I've added caveats to this and its closely related functions, such as:
`Note that on error new elements may remain added to req if req != NULL.`